### PR TITLE
Feat/#880_ GET and DELETE improved error handling

### PR DESCRIPTION
**Pull request by ciaranschutte**
**Description of changes**
- stop throwing errors all the way up the stack, do explicit falsey check for returned documents. no docs found will not throw an error but return [] or null depending on query.
- better error handling, responses and messages
- add middle layer for GET and DELETE requests for uniformity and cleaner distinction of responsibilities (error message comes from "service" file)
- return lean mongo document directly from api instead of using our util
- use `undefined` consistently instead of `nulls`
- adds correct http responses

**Type of Change**

- [ ] Bug
- [ ] Refactor
- [ ] New Feature
- [ ] Release Candidate

**Checklist before requesting review:**

- [ ] Check branch (code change PRs go to `develop` not master)
- [ ] Check copyrights for new files
- [ ] Manual testing
- [ ] Regression tests completed and passing (double check number of tests).
- [ ] Spelling has been checked.
- [ ] Updated swagger docs accordingly (check it's still valid)
- [ ] Set `validationDependency` in meta tag for [Argo Dictionary](https://github.com/icgc-argo/argo-dictionary) fields used in code

### CODE REVIEW

**Comment by ciaranschutte** [comment]
pending build issues with eslint

related PR to change eslint rules: https://github.com/icgc-argo/argo-clinical/pull/892

File: src/exception/exception-api.ts [hunk 1]
```diff
@@ -77,15 +65,20 @@ class ExceptionController {
     }
   }
 
+  @HasFullWriteAccess()
+  async clearProgramException(req: Request, res: Response) {
+    const programId = req.params.programId;
+    const result = await exceptionService.operations.deleteProgramException({ programId });
```
Comment (caravinci) [inline src/exception/exception-api.ts:71]: While I like the DRY approach you went for, the `try/catch` in the original implementation ensures catching any other possible errors within `...deleteProgramException`.
perhaps you can wrap this in it again, and add respond with a 500 if there's any issues?

File: src/exception/exception-api.ts [hunk 2]
```diff
@@ -77,15 +65,20 @@ class ExceptionController {
...
     const programId = req.params.programId;
-    try {
-      const exception = await programExceptionRepository.find(programId);
-      res.status(200).send(exception);
-    } catch (e) {
-      res.status(200).send({});
-    }
+    const result = await exceptionService.operations.getProgramException({ programId });
+    const status = result.success ? 200 : 400;
+    return res.status(status).send(result);
```
Comment (caravinci) [inline src/exception/exception-api.ts:81]: same here re `try/catch`. an additional 500 would be a good improvement

File: src/exception/exception-repo.ts [hunk 3]
```diff
@@ -35,58 +34,61 @@ const programExceptionSchema = new mongoose.Schema({
...
 
-type ProgramExceptionDocument = mongoose.Document & ProgramException;
-
-export const ProgramExceptionModel = mongoose.model<ProgramExceptionDocument>(
+export const ProgramExceptionModel = mongoose.model<ProgramException>(
   'ProgramException',
   programExceptionSchema,
 );
 
+type RepoResponse = Promise<ProgramException | undefined>;
```
Comment (caravinci) [inline src/exception/exception-repo.ts:42]: 🎉 I like this response unification

File: src/exception/exception-repo.ts [hunk 4]
```diff
@@ -35,58 +34,61 @@ const programExceptionSchema = new mongoose.Schema({
...
-  find(programId: string): Promise<DeepReadonly<ProgramException> | undefined>;
-  delete(programId: string): Promise<void>;
+  save(exception: ProgramException): RepoResponse;
+  find(programId: string): RepoResponse;
+  delete(programId: string): RepoResponse;
 }
 
+const checkDoc = (doc: null | ProgramException): undefined | ProgramException => {
+  if (doc === null || (isArray(doc) && doc.length === 0)) {
+    return undefined;
```
Comment (caravinci) [inline src/exception/exception-repo.ts:52]: no need to change it, as this works perfectly, it caught me by surprise like with the "automatic semicolon insertion" battles, I didn't realise I'm more inclined towards `return;` than explicitly returning an `undefined`.
🤷 
Now I know.

**Review by caravinci — APPROVED** [review]
💪 ship it!

**Pull request merged**